registry/storage/driver/s3-aws/s3.go

```diff
@@ -1082,110 +1083,87 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
+	)
+	prevDir = prefix + path
 
 	listObjectsInput := &s3.ListObjectsV2Input{
-		Bucket:    aws.String(d.Bucket),
-		Prefix:    aws.String(path),
-		Delimiter: aws.String("/"),
-		MaxKeys:   aws.Int64(listMax),
+		Bucket:  aws.String(d.Bucket),
+		Prefix:  aws.String(path),
+		MaxKeys: aws.Int64(listMax),
 	}
 
 	s := d.s3Client(parentCtx)
 
 	ctx, done := dcontext.WithTrace(parentCtx)
 	defer done("s3aws.ListObjectsV2Pages(%s)", path)
-	listObjectErr := s.ListObjectsV2PagesWithContext(ctx, listObjectsInput, func(objects *s3.ListObjectsV2Output, lastPage bool) bool {
 
-		var count int64
-		// KeyCount was introduced with version 2 of the GET Bucket operation in S3.
-		// Some s3 implementations (looking at you ceph/rgw) have a buggy
-		// implementation so we intionally avoid ever using it, preferring instead
-		// to calculate the count from the Contents and CommonPrefixes fields of
-		// the s3.ListObjectsV2Output. we retain the commented out KeyCount code
-		// and this comment so as not to forget this problem moving forward.
-		//
-		// count = *objects.KeyCount
-		// *objectCount += *objects.KeyCount
-
-		count = int64(len(objects.Contents) + len(objects.CommonPrefixes))
-		*objectCount += count
-
-		walkInfos := make([]walkInfoContainer, 0, count)
-
-		for _, dir := range objects.CommonPrefixes {
-			commonPrefix := *dir.Prefix
-			walkInfos = append(walkInfos, walkInfoContainer{
-				prefix: dir.Prefix,
-				FileInfoFields: storagedriver.FileInfoFields{
-					IsDir: true,
-					Path:  strings.Replace(commonPrefix[:len(commonPrefix)-1], d.s3Path(""), prefix, 1),
-				},
-			})
-		}
+	// The s3 list object API will list all objects under the starting path recursively, in sorted,
+	// depth-first ordering https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html
+	// S3 does not return directory without the "delimiter":"/" argument passed, but they can to be
+	// inferred from object paths and then de-duplicated to avoid calling the walk fn redundantly.
+	// With files returned in sorted depth-first order, directories are inferred in the same order,
+	// so avoiding redundant directories is a matter of not passing the same directory to the walk fn twice in a row.
+	// ErrSkipDir is handled by explicitly skipping over any files under the skipped directory. This may sub-optimal
+	// for extreme edge cases but for the general use case in a registry, this is orders of magnitude
+	// faster than a more explicit recursive implementation.
+	listObjectErr := s.ListObjectsV2PagesWithContext(ctx, listObjectsInput, func(objects *s3.ListObjectsV2Output, lastPage bool) bool {
+		walkInfos := make([]storagedriver.FileInfoInternal, 0, len(objects.Contents))
 
 		for _, file := range objects.Contents {
-			walkInfos = append(walkInfos, walkInfoContainer{
+			filePath := strings.Replace(*file.Key, d.s3Path(""), prefix, 1)
+
+			// get a list of all inferred directories skipped between the previous directory and this file
```

Comment:
Nit: "skipped" is confusing, since we're not implementing skipping in the `ErrSkipDir` sense here. I think the comment would be clear if you just omit the word "skipped" - we're finding all the inferred directories.